### PR TITLE
Fix goto-cc when multiple files are given

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0b464aed4a77438992f5f8bf0bab332a8286e806).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79909056

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0546c72f69d945995b9755fda2a2773d954a3129).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79953154

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: c5115c44eb59b192d75b3039bdc06bfaa57c05af).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79986984
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 52925c6ec1962bdf19b4f3d0003d3671bd618557).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79993217
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

File: src/goto-cc/compile.cpp [hunk 1]
```diff
@@ -391,8 +391,9 @@ bool compilet::compile()
     source_files.pop_front();
 
     // Visual Studio always prints the name of the file it's doing
+    // onto stdout. The name of the directory is stripped.
     if(echo_file_name)
-      status() << file_name << eom;
+      std::cout << get_base_name(file_name, false) << '\n' << std::flush;
```
Comment (smowton) [inline src/goto-cc/compile.cpp:396]: Guessing this is only done for want of a `flush()` method on `messaget`, as I guess our driver knows that the messaget needs to point at stdout?

File: src/goto-cc/compile.h [hunk 2]
```diff
@@ -45,7 +45,8 @@ class compilet:public language_uit
   std::list<irep_idt> seen_modes;
 
   std::string object_file_extension;
-  std::string output_file_object, output_file_executable;
+  std::string output_file_object, output_directory_object;
```
Comment (smowton) [inline src/goto-cc/compile.h]: Comment that these are mutually exclusive, and are not a pathname / basename pair as they appear

File: src/util/get_base_name.cpp [hunk 3]
```diff
@@ -8,31 +8,39 @@ Author: CM Wintersteiger
...
 std::string get_base_name(const std::string &in, bool strip_suffix)
 {
-  size_t r=std::string::npos;
-  if(strip_suffix)
-    r=in.rfind('.', in.length()-1);
-  if(r==std::string::npos)
-    r=in.length();
+#ifdef _WIN32
+  // Windows now allows both '/' and '\\'
+  const std::size_t dash_pos = in.find_last_of("\\/");
```
Comment (smowton) [inline src/util/get_base_name.cpp]: `slash_pos`

File: src/util/get_base_name.cpp [hunk 4]
```diff
@@ -8,31 +8,39 @@ Author: CM Wintersteiger
...
 
-  f = (fw>f)?fw:f;
+  std::size_t char_count;
 
-  if(in[f]=='/' || in[f]=='\\')
-    f++;
-  return in.substr(f, r-f);
+  if(strip_suffix)
+  {
+    std::size_t dot_pos = in.rfind('.');
```
Comment (smowton) [inline src/util/get_base_name.cpp:36]: `dot_pos` could be `npos`?

File: src/goto-cc/ms_link_cmdline.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+  for(int i = 1; i < argc; i++)
+    options.push_back(argv[i]);
+
+  return parse(options);
+}
+
+static std::istream &my_wgetline(std::istream &in, std::wstring &dest)
+{
+  // We should support this properly,
+  // but will just strip right now.
```
Comment (smowton) [inline src/goto-cc/ms_link_cmdline.cpp:80]: ? It seems like we're straightforwardly packing chars instead?

File: src/goto-cc/ms_link_cmdline.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+  parse(options);
+}
+
+/// \return none
+void ms_link_cmdlinet::process_non_link_option(const std::string &s)
+{
+  set(s);
+
+  for(unsigned j = 0; non_ms_link_options[j] != nullptr; j++)
```
Comment (smowton) [inline src/goto-cc/ms_link_cmdline.cpp]: Consider c++11 for-each loop instead of null terminator?

File: src/goto-cc/ms_link_cmdline.cpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+  if(s == "")
+    return;
+
+  if(s[0] != '/' && s[0] != '-')
+  {
+    args.push_back(s);
+    return;
+  }
+
+  for(std::size_t j = 0; ms_link_options[j] != nullptr; j++)
```
Comment (smowton) [inline src/goto-cc/ms_link_cmdline.cpp]: Same here, no need for the magic terminator

File: src/goto-cc/ms_link_cmdline.cpp [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+  {
+    const std::string ms_link_option = ms_link_options[j];
+
+    if(
+      std::string(s, 1, std::string::npos) == ms_link_option ||
```
Comment (smowton) [inline src/goto-cc/ms_link_cmdline.cpp]: Use `.compare` for the same thing without copying `s`?

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-cc/compile.cpp:396]: No, this is really meant to ensure that the message goes to stdout, irrespectively of level settings etc

File: src/goto-cc/compile.h (hunk 2 above)
Comment (kroening) [reply src/goto-cc/compile.h]: done

File: src/util/get_base_name.cpp (hunk 3 above)
Comment (kroening) [reply src/util/get_base_name.cpp]: done

File: src/util/get_base_name.cpp (hunk 4 above)
Comment (kroening) [reply src/util/get_base_name.cpp:36]: Indeed, fixed

File: src/goto-cc/ms_link_cmdline.cpp (hunk 6 above)
Comment (kroening) [reply src/goto-cc/ms_link_cmdline.cpp]: done

File: src/goto-cc/ms_link_cmdline.cpp (hunk 7 above)
Comment (kroening) [reply src/goto-cc/ms_link_cmdline.cpp]: done

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (thk123) [reply src/goto-cc/compile.cpp:396]: Excuse me for wading in here, but this essentially breaks `--json-ui` - don't know if that's a concern here but makes it harder to use this from a calling program. May at least want to warn if json ui is enabled on the message_handler?

File: src/goto-cc/ms_link_cmdline.cpp (hunk 8 above)
Comment (kroening) [reply src/goto-cc/ms_link_cmdline.cpp]: yes, but won't work for the second case, with the ':'

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-cc/compile.cpp:396]: goto-cc does not have the json ui option.

Comment (smowton) [reply src/goto-cc/compile.cpp:396]: Ah ok, so maybe just add a highest prio level to `messaget`? Or an alias for `error()` if that can never be silenced even with `--verbosity 0`?

**Comment by smowton** [comment]
Looks good except I think we should do `messaget::always()` or whatever instead of grabbing `std::cout` and surprising anyone that's trying to do message filtering or tee'ing or whatever

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 408dffac44824b71c0d65c2213963f984ee238cb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80006533
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8db6d479d9f0d9cb1d0a5a00d85b96d0973148f2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80007670
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a466de7a9d71e784d088f4c6d92e5b5fdf78a440).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80010020

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-cc/compile.cpp:396]: No, won't work -- the spec says to output to stdout, and error() will output to stderr.

File: src/goto-cc/compile.cpp [hunk 9]
```diff
@@ -421,9 +421,10 @@ bool compilet::compile()
       std::string cfn;
 
       if(output_file_object=="")
-        cfn=get_base_name(file_name, true)+"."+object_file_extension;
+        cfn = output_directory_object + get_base_name(file_name, true) + "." +
```
Comment (tautschnig) [inline src/goto-cc/compile.cpp]: I believe this lacks a path separator after `output_directory_object` (which should be appended to `output_directory_object` if it is non-empty).

**Review by tautschnig — CHANGES_REQUESTED** [review]
I believe one fix is required as noted below; otherwise the absence of tests makes me feel uneasy. I think this should have been several PRs with tests for each fix. But I'm not going to block on those.

File: src/goto-cc/compile.cpp (hunk 9 above)
Comment (kroening) [reply src/goto-cc/compile.cpp]: done

**Comment by kroening** [comment]
For tests, we would need to open up a new set -- say regression/goto-cl.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e765dbc37368cdb995dba950be85abb3d65c7de2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80424928

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-cc/compile.cpp:396]: How about using the `cout_message_handlert`?

**Comment by tautschnig** [comment]
> For tests, we would need to open up a new set -- say regression/goto-cl.

Yes, indeed the case for the last three commits. The first two, however, should be testable as is. (Notably, "fix get_base_name" is a code change without explanation what the bug was - a test would provide such an explanation.)

**Comment by kroening** [comment]
Split into pieces

**Pull request closed without merging**